# Diving deep into batting within Baseball
## Authors: Adan Hernandez & Isaiah Dawkins

### Dataset Identification and Acquisition
For this project, we have selected a dataset containing Major League Baseball (MLB) batting statistics for all players during the 2024 and 2025 seasons. Baseball is a data-rich sport, and batting statistics offer numerous insights into player performance, league-wide trends, and team strategies. The data was sourced from [Baseball Reference](https://www.baseball-reference.com/leagues/majors/2025-standard-batting.shtml#players_standard_batting), a widely recognized and reliable platform for professional baseball statistics.

The dataset includes player-level stats such as:
- **Player information**: name, age, team, position, and handedness (where applicable)
- **Batting performance metrics**: games played (G), at-bats (AB), hits (H), doubles (2B), triples (3B), home runs (HR), runs batted in (RBI), walks (BB), strikeouts (SO), batting average (AVG), on-base percentage (OBP), slugging percentage (SLG), on-base plus slugging (OPS), and much more
- **Seasonal context**: year (will be added), team, and league (AL/NL)

The dataset was downloaded as CSV files — one for each season (2024 and 2025). Each file contains one row per player per season, including both qualified hitters and those with limited plate appearances.

##### Some of the questions we aim to answer are:
1. **Seasonal Comparison**:
   - How did overall league batting performance (e.g., AVG, HR, OPS) differ between 2024 and 2025?
2. **Top Performers**:
   - Which players showed the greatest improvement or decline in offensive metrics between the two seasons?
3. **Power vs. Contact**:
   - Is there a clear tradeoff between power hitters (HR leaders) and batting average consistency?
4. **Team-Level Analysis**:
   - Which teams had the most productive lineups, and how did their team batting stats correlate with win totals?
5. **Player Efficiency**:
   - Are high on-base percentages (OBP) more strongly correlated with runs scored (R) or RBIs?

In [77]:
import pandas as pd

### Importing our Data Sets

In [78]:
batting_25_df = pd.read_csv('./Batting_2025.csv')
batting_25_df.head(10)

,Rk,Player,Age,Team,Lg,WAR,G,PA,AB,R,...,Rbat+,TB,GIDP,HBP,SH,SF,IBB,Pos,Awards,Player-additional
0,1,Francisco Lindor#,31,NYM,NL,5.8,160,732,644,117,...,128.0,300,9,16,0,7,2,*6/DH,AS,lindofr01
1,2,Rafael Devers*,28,2TM,2LG,4.0,163,729,607,99,...,139.0,291,16,6,0,4,10,*D3/5H,NaN,deverra01
2,2,Rafael Devers*,28,BOS,AL,2.3,73,334,272,47,...,153.0,137,7,4,0,2,7,D,NaN,deverra01
3,2,Rafael Devers*,28,SFG,NL,1.8,90,395,335,52,...,127.0,154,9,2,0,2,3,D3/5H,NaN,deverra01
4,3,Shohei Ohtani*,30,LAD,NL,6.6,158,727,611,146,...,175.0,380,9,3,0,2,20,*D1,AS,ohtansh01
5,4,Matt Olson*,31,ATL,NL,6.1,162,724,624,98,...,136.0,302,14,4,0,5,2,*3,AS,olsonma02
6,5,Kyle Schwarber*,32,PHI,NL,4.7,162,724,604,111,...,149.0,340,4,11,0,1,6,*D/7,AS,schwaky01
7,6,Geraldo Perdomo#,25,ARI,NL,7.0,161,720,597,98,...,139.0,276,11,10,8,11,1,*6/H,NaN,perdoge01
8,7,Juan Soto*,26,NYM,NL,6.2,160,715,577,120,...,163.0,303,17,3,1,6,13,*9/D,NaN,sotoju01
9,8,Julio Rodríguez,24,SEA,AL,6.8,160,710,652,106,...,133.0,309,15,12,0,2,5,*8/DH,AS,rodriju01


In [79]:
batting_24_df = pd.read_csv('./Batting_2024.csv')
batting_24_df.head(10)

,Rk,Player,Age,Team,Lg,WAR,G,PA,AB,R,...,Rbat+,TB,GIDP,HBP,SH,SF,IBB,Pos,Awards,Player-additional
0,1,Jarren Duran*,27,BOS,AL,8.7,160,735,671,111,...,134.0,330,6,6,1,3,1,*87,ASMVP-8,duranja01
1,2,Shohei Ohtani*,29,LAD,NL,9.2,159,731,636,134,...,190.0,411,7,6,0,5,10,*D,ASMVP-1SS,ohtansh01
2,3,Gunnar Henderson*,23,BAL,AL,9.1,159,719,630,118,...,157.0,333,2,7,0,4,1,*6/D,ASMVP-4,hendegu01
3,4,Marcus Semien,33,TEX,AL,4.1,159,718,650,101,...,100.0,254,9,3,0,1,2,*4,AS,semiema01
4,5,Juan Soto*,25,NYY,AL,7.9,157,713,576,128,...,179.0,328,10,4,0,4,2,*9/7DH,ASMVP-3SS,sotoju01
5,6,Bobby Witt Jr.,24,KCR,AL,9.4,161,709,636,125,...,171.0,374,4,8,0,8,9,*6/D,ASMVP-2GGSS,wittbo02
6,7,Aaron Judge,32,NYY,AL,10.8,158,704,559,122,...,223.0,392,22,9,0,2,20,*8D/97,ASMVP-1SS,judgeaa01
7,8,Vladimir Guerrero Jr.,25,TOR,AL,6.2,159,697,616,98,...,165.0,335,16,5,0,4,12,*3D5/H,ASMVP-6SS,guerrvl02
8,9,Elly De La Cruz#,22,CIN,NL,5.2,160,696,618,105,...,122.0,291,12,6,2,1,2,*6/H,ASMVP-8,delacel01
9,10,Pete Alonso,29,NYM,NL,2.6,162,695,608,91,...,122.0,279,14,13,0,4,4,*3/DH,AS,alonspe01


### Combine our data into one CSV/dataset
#### Step 0.) Combine our two data sets from 2024 & 2025 to make a larger single data set for ease of use and cleaning process

In [80]:
# Add column to our tables
batting_24_df['Year'] = "2024"
batting_25_df['Year'] = "2025"

# Concat the two csv's and save it
batting_data_df = pd.concat([batting_24_df, batting_25_df], ignore_index=True)
batting_data_df.to_csv('batting_data_24_25.csv', index=False)

# Sanity check to make sure we have one dataset
batting_data_df.head(10)

,Rk,Player,Age,Team,Lg,WAR,G,PA,AB,R,...,TB,GIDP,HBP,SH,SF,IBB,Pos,Awards,Player-additional,Year
0,1,Jarren Duran*,27,BOS,AL,8.7,160,735,671,111,...,330,6,6,1,3,1,*87,ASMVP-8,duranja01,2024
1,2,Shohei Ohtani*,29,LAD,NL,9.2,159,731,636,134,...,411,7,6,0,5,10,*D,ASMVP-1SS,ohtansh01,2024
2,3,Gunnar Henderson*,23,BAL,AL,9.1,159,719,630,118,...,333,2,7,0,4,1,*6/D,ASMVP-4,hendegu01,2024
3,4,Marcus Semien,33,TEX,AL,4.1,159,718,650,101,...,254,9,3,0,1,2,*4,AS,semiema01,2024
4,5,Juan Soto*,25,NYY,AL,7.9,157,713,576,128,...,328,10,4,0,4,2,*9/7DH,ASMVP-3SS,sotoju01,2024
5,6,Bobby Witt Jr.,24,KCR,AL,9.4,161,709,636,125,...,374,4,8,0,8,9,*6/D,ASMVP-2GGSS,wittbo02,2024
6,7,Aaron Judge,32,NYY,AL,10.8,158,704,559,122,...,392,22,9,0,2,20,*8D/97,ASMVP-1SS,judgeaa01,2024
7,8,Vladimir Guerrero Jr.,25,TOR,AL,6.2,159,697,616,98,...,335,16,5,0,4,12,*3D5/H,ASMVP-6SS,guerrvl02,2024
8,9,Elly De La Cruz#,22,CIN,NL,5.2,160,696,618,105,...,291,12,6,2,1,2,*6/H,ASMVP-8,delacel01,2024
9,10,Pete Alonso,29,NYM,NL,2.6,162,695,608,91,...,279,14,13,0,4,4,*3/DH,AS,alonspe01,2024


### Data Cleaning and Preparation

#### Step 1.) Add a column for handedness, map handedness values. 
##### Our data was imported with characters alongside Player names, indicating batter handedness. 
##### * is left handed, # both handedness, nothing is right handed, ? is unknown

In [93]:
# Add column for handedness
batting_data_df['Handedness'] = "Right"

# Masks for checking the handedness, as shown on the Player name before cleaning
left_mask = batting_data_df['Player'].str.contains('*', regex=False)
both_mask = batting_data_df['Player'].str.contains('#', regex=False)
unknown_mask = batting_data_df['Player'].str.contains('?', regex=False)

# Set the appropriate handedness after we parse for each type
batting_data_df.loc[left_mask, 'Handedness'] = 'Left'
batting_data_df.loc[both_mask, 'Handedness'] = 'Both'
batting_data_df.loc[unknown_mask, 'Handedness'] = 'Unknown'

batting_data_df.head(10)

,Rk,Player,Age,Team,Lg,WAR,G,PA,AB,R,...,GIDP,HBP,SH,SF,IBB,Pos,Awards,Player-additional,Year,Handedness
1717,664,Greg Jones,27,CHW,AL,0.0,3,2,2,0,...,0,0,0,0,0,/H9,None,jonesgr02,2025,Right
1718,665,Chuckie Robinson,30,LAD,NL,0.0,1,2,1,1,...,0,0,1,0,0,/2,None,robinch04,2025,Right
1719,666,Trevor Rogers,27,BAL,AL,0.0,1,2,2,0,...,0,0,0,0,0,1,None,rogertr01,2025,Right
1720,667,Tyler Alexander,30,CHW,AL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,alexaty01,2025,Right
1721,668,Jack Leiter,25,TEX,AL,0.0,1,1,1,0,...,0,0,0,0,0,1/H,None,leiteja01,2025,Right
1722,669,Ryne Nelson,27,ARI,NL,0.1,1,1,1,1,...,0,0,0,0,0,1/H,None,nelsory01,2025,Right
1723,670,Shinnosuke Ogasawara,27,WSN,NL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,ogasash01,2025,Right
1724,671,Colin Rea,34,CHC,NL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,reaco01,2025,Right
1725,672,Garrett Stubbs,32,PHI,NL,0.0,5,1,1,2,...,0,0,0,0,0,/DH2,None,stubbga01,2025,Right
1726,673,Terrin Vavra,28,BAL,AL,0.0,1,1,1,0,...,0,0,0,0,0,/4H,None,vavrate01,2025,Right


#### Step 2.) Remove flags that were added to Player names
Since the flags have been properly mapped to the correct values in the Handedness column, we want to remove these characters

In [91]:
batting_data_df.loc[left_mask, 'Player'] = batting_data_df['Player'].str[:-1]
batting_data_df.loc[both_mask, 'Player'] = batting_data_df['Player'].str[:-1]
batting_data_df.loc[unknown_mask, 'Player'] = batting_data_df['Player'].str[:-1]

batting_data_df.head(10)

,Rk,Player,Age,Team,Lg,WAR,G,PA,AB,R,...,GIDP,HBP,SH,SF,IBB,Pos,Awards,Player-additional,Year,Handedness
1717,664,Greg Jones,27,CHW,AL,0.0,3,2,2,0,...,0,0,0,0,0,/H9,None,jonesgr02,2025,Both
1718,665,Chuckie Robinson,30,LAD,NL,0.0,1,2,1,1,...,0,0,1,0,0,/2,None,robinch04,2025,Right
1719,666,Trevor Rogers,27,BAL,AL,0.0,1,2,2,0,...,0,0,0,0,0,1,None,rogertr01,2025,Left
1720,667,Tyler Alexander,30,CHW,AL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,alexaty01,2025,Right
1721,668,Jack Leiter,25,TEX,AL,0.0,1,1,1,0,...,0,0,0,0,0,1/H,None,leiteja01,2025,Right
1722,669,Ryne Nelson,27,ARI,NL,0.1,1,1,1,1,...,0,0,0,0,0,1/H,None,nelsory01,2025,Right
1723,670,Shinnosuke Ogasawara,27,WSN,NL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,ogasash01,2025,Left
1724,671,Colin Rea,34,CHC,NL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,reaco01,2025,Right
1725,672,Garrett Stubbs,32,PHI,NL,0.0,5,1,1,2,...,0,0,0,0,0,/DH2,None,stubbga01,2025,Left
1726,673,Terrin Vavra,28,BAL,AL,0.0,1,1,1,0,...,0,0,0,0,0,/4H,None,vavrate01,2025,Left


#### We must do the same for the Pos column

In [90]:
# Reassigning the masks to check for special characters in the Pos column
left_mask = batting_data_df['Pos'].str.contains('*', regex=False)
both_mask = batting_data_df['Pos'].str.contains('#', regex=False)
unknown_mask = batting_data_df['Pos'].str.contains('?', regex=False)

# Removing the special characters in the Pos column
batting_data_df.loc[left_mask, 'Pos'] = batting_data_df['Pos'].str.replace('*', '', regex=False)
batting_data_df.loc[both_mask, 'Pos'] = batting_data_df['Pos'].str.replace('#', '', regex=False)
batting_data_df.loc[unknown_mask, 'Pos'] = batting_data_df['Pos'].str.replace('?', '', regex=False)

batting_data_df.head(10)

,Rk,Player,Age,Team,Lg,WAR,G,PA,AB,R,...,GIDP,HBP,SH,SF,IBB,Pos,Awards,Player-additional,Year,Handedness
1717,664,Greg Jones,27,CHW,AL,0.0,3,2,2,0,...,0,0,0,0,0,/H9,None,jonesgr02,2025,Both
1718,665,Chuckie Robinson,30,LAD,NL,0.0,1,2,1,1,...,0,0,1,0,0,/2,None,robinch04,2025,Right
1719,666,Trevor Rogers,27,BAL,AL,0.0,1,2,2,0,...,0,0,0,0,0,1,None,rogertr01,2025,Left
1720,667,Tyler Alexander,30,CHW,AL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,alexaty01,2025,Right
1721,668,Jack Leiter,25,TEX,AL,0.0,1,1,1,0,...,0,0,0,0,0,1/H,None,leiteja01,2025,Right
1722,669,Ryne Nelson,27,ARI,NL,0.1,1,1,1,1,...,0,0,0,0,0,1/H,None,nelsory01,2025,Right
1723,670,Shinnosuke Ogasawara,27,WSN,NL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,ogasash01,2025,Left
1724,671,Colin Rea,34,CHC,NL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,reaco01,2025,Right
1725,672,Garrett Stubbs,32,PHI,NL,0.0,5,1,1,2,...,0,0,0,0,0,/DH2,None,stubbga01,2025,Left
1726,673,Terrin Vavra,28,BAL,AL,0.0,1,1,1,0,...,0,0,0,0,0,/4H,None,vavrate01,2025,Left


#### Step 3.) Check the info of the data frame for any null values

In [84]:
batting_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1817 entries, 0 to 1816
Data columns (total 37 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Rk                 1817 non-null   int64  
 1   Player             1817 non-null   object 
 2   Age                1817 non-null   int64  
 3   Team               1817 non-null   object 
 4   Lg                 1817 non-null   object 
 5   WAR                1815 non-null   float64
 6   G                  1817 non-null   int64  
 7   PA                 1817 non-null   int64  
 8   AB                 1817 non-null   int64  
 9   R                  1817 non-null   int64  
 10  H                  1817 non-null   int64  
 11  2B                 1817 non-null   int64  
 12  3B                 1817 non-null   int64  
 13  HR                 1817 non-null   int64  
 14  RBI                1817 non-null   int64  
 15  SB                 1817 non-null   int64  
 16  CS                 1817 

#### The Awards column seems to have a considerable amount of NULL values.
#### We will first analyze the column to have a better understanding of how to handle the NULL values.

In [85]:
batting_data_df['Awards'].head(30)

0         ASMVP-8
1       ASMVP-1SS
2         ASMVP-4
3              AS
4       ASMVP-3SS
5     ASMVP-2GGSS
6       ASMVP-1SS
7       ASMVP-6SS
8         ASMVP-8
9              AS
10       MVP-19GG
11             AS
12         MVP-15
13        MVP-2SS
14            NaN
15         MVP-10
16        ASMVP-4
17            NaN
18            NaN
19     ASMVP-15SS
20      ASMVP-5SS
21      ASMVP-5SS
22       ASMVP-18
23            NaN
24            NaN
25     ASMVP-14SS
26     ASMVP-14SS
27            NaN
28            NaN
29            NaN
Name: Awards, dtype: object

#### The Awards column lists any awards the player has won for the season.
#### We can treat the NULL values as having no awards and use fillna() to convert the NULL types to 'None'.

In [86]:
batting_data_df = batting_data_df.fillna({'Awards': 'None'})

batting_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1817 entries, 0 to 1816
Data columns (total 37 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Rk                 1817 non-null   int64  
 1   Player             1817 non-null   object 
 2   Age                1817 non-null   int64  
 3   Team               1817 non-null   object 
 4   Lg                 1817 non-null   object 
 5   WAR                1815 non-null   float64
 6   G                  1817 non-null   int64  
 7   PA                 1817 non-null   int64  
 8   AB                 1817 non-null   int64  
 9   R                  1817 non-null   int64  
 10  H                  1817 non-null   int64  
 11  2B                 1817 non-null   int64  
 12  3B                 1817 non-null   int64  
 13  HR                 1817 non-null   int64  
 14  RBI                1817 non-null   int64  
 15  SB                 1817 non-null   int64  
 16  CS                 1817 

### The awards column has no null values now.
### Columns BA, OBP, SLG, OPS, OPS+, rOBA, and Rbat+ have roughly 180 NULL values

We will first analyze the data we get from these rows containing null values. We will be looking at the those columns as well as the AB (at-bats) column to see if those players have any data to provide in the first place.

In [87]:
# Creating a mask to find all the null values in BA column
ba_null_mask = batting_data_df['BA'].isnull()

# Creating a list to indicate columns to select in next step
na_cols = ['AB','BA', 'OBP', 'SLG', 'OPS', 'OPS+', 'rOBA', 'Rbat+']

# Selecting all the rows with null values in BA
ba_null_df = batting_data_df[ba_null_mask]

ba_null_df[na_cols].head(30)

,AB,BA,OBP,SLG,OPS,OPS+,rOBA,Rbat+
636,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
716,0,NaN,1.0,NaN,NaN,NaN,0.691,373.0
803,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
804,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
805,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
806,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
807,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
808,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
809,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
810,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Since the rows that have null values in these columns signify players that had zero at-bats, we can drop them from the data.

In [88]:
batting_data_df = batting_data_df.dropna()

batting_data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1629 entries, 0 to 1726
Data columns (total 37 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Rk                 1629 non-null   int64  
 1   Player             1629 non-null   object 
 2   Age                1629 non-null   int64  
 3   Team               1629 non-null   object 
 4   Lg                 1629 non-null   object 
 5   WAR                1629 non-null   float64
 6   G                  1629 non-null   int64  
 7   PA                 1629 non-null   int64  
 8   AB                 1629 non-null   int64  
 9   R                  1629 non-null   int64  
 10  H                  1629 non-null   int64  
 11  2B                 1629 non-null   int64  
 12  3B                 1629 non-null   int64  
 13  HR                 1629 non-null   int64  
 14  RBI                1629 non-null   int64  
 15  SB                 1629 non-null   int64  
 16  CS                 1629 non-n

#### We now have a data frame of 1629 rows, with no null values, correct data types, and cleaned values without special characters

In [89]:
batting_data_df

,Rk,Player,Age,Team,Lg,WAR,G,PA,AB,R,...,GIDP,HBP,SH,SF,IBB,Pos,Awards,Player-additional,Year,Handedness
0,1,Jarren Duran,27,BOS,AL,8.7,160,735,671,111,...,6,6,1,3,1,87,ASMVP-8,duranja01,2024,Left
1,2,Shohei Ohtani,29,LAD,NL,9.2,159,731,636,134,...,7,6,0,5,10,D,ASMVP-1SS,ohtansh01,2024,Left
2,3,Gunnar Henderson,23,BAL,AL,9.1,159,719,630,118,...,2,7,0,4,1,6/D,ASMVP-4,hendegu01,2024,Left
3,4,Marcus Semien,33,TEX,AL,4.1,159,718,650,101,...,9,3,0,1,2,4,AS,semiema01,2024,Right
4,5,Juan Soto,25,NYY,AL,7.9,157,713,576,128,...,10,4,0,4,2,9/7DH,ASMVP-3SS,sotoju01,2024,Left
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1722,669,Ryne Nelson,27,ARI,NL,0.1,1,1,1,1,...,0,0,0,0,0,1/H,None,nelsory01,2025,Right
1723,670,Shinnosuke Ogasawara,27,WSN,NL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,ogasash01,2025,Left
1724,671,Colin Rea,34,CHC,NL,0.0,1,1,1,0,...,0,0,0,0,0,1,None,reaco01,2025,Right
1725,672,Garrett Stubbs,32,PHI,NL,0.0,5,1,1,2,...,0,0,0,0,0,/DH2,None,stubbga01,2025,Left
